File: tools/bench_fa2_profile.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <set>
#include <tuple>
#include <sstream>
#include <vector>
#include <cmath>
#include <iomanip>
#include <filesystem>
#include "flash_attn/attn.hpp"
#include "flash_attn/io.hpp"
#include "flash_attn/args.hpp"

int main(int argc, char** argv) {
    const int T            = arg_int(argc, argv, "--T", 1024);
    const int d            = arg_int(argc, argv, "--d", 64);
    const int M_bytes      = arg_int(argc, argv, "--M_bytes", 131072);
    const int num_testsets = arg_int(argc, argv, "--num_testsets", 100);
    const std::string opt  = arg_str(argc, argv, "--opt_flag", "O3");

    const std::string out_dir  = "outputs/fa2_cpp_profile_" + opt;
    const std::string log_path = out_dir + "/runtime.csv";
    std::filesystem::create_directories(out_dir);

    std::set<std::tuple<int,int,int,int>> done;
    if (std::filesystem::exists(log_path)) {
        std::ifstream fin(log_path);
        std::string line;
        std::getline(fin, line);
        while (std::getline(fin, line)) {
            std::stringstream ss(line);
            std::string seg;
            std::vector<std::string> p;
            while (std::getline(ss, seg, ',')) p.push_back(seg);
            if (p.size() >= 4) {
                try { done.insert({std::stoi(p[0]), std::stoi(p[1]),
                                   std::stoi(p[2]), std::stoi(p[3])}); }
                catch (...) {}
            }
        }
    } else {
        std::ofstream f(log_path);
        f << "testset,T,d,M_bytes,runtime,line7,line8,line10,line11,others\n";
    }

    const float scale = 1.0f / std::sqrt(float(d));

    for (int i = 0; i < num_testsets; ++i) {
        if (done.count({i, T, d, M_bytes})) {
            std::cout << "Skipping testset " << i << "\n";
            continue;
        }

        const std::string data_dir = "data/testset" + std::to_string(i);
        auto Q  = load_matrix(data_dir + "/Q.txt",  T, d);
        auto Kt = load_matrix(data_dir + "/Kt.txt", d, T);
        auto V  = load_matrix(data_dir + "/V.txt",  T, d);

        auto [O, p] = fa2_forward_profile(Q, Kt, V, T, d, M_bytes, scale);

        std::cout << "---------------------------------------\n"
                  << "Runtime " << i << ": " << std::fixed << std::setprecision(4)
                  << p.runtime << " s\n"
                  << "Line 7: "  << p.t7   << " s  "
                  << "Line 8: "  << p.t8   << " s  "
                  << "Line 10: " << p.t10  << " s  "
                  << "Line 11: " << p.t11  << " s  "
                  << "Others: "  << p.others << " s\n";

        std::ofstream f(log_path, std::ios::app);
        f << i << "," << T << "," << d << "," << M_bytes << ","
          << std::fixed << std::setprecision(6)
          << p.runtime << "," << p.t7 << "," << p.t8 << ","
          << p.t10 << "," << p.t11 << "," << p.others << "\n";
    }
    return 0;
}

File: tools/bench_fa2_profile.py
import sys
import os
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np
import argparse
from flash_attn.attn_profile import fa2_forward_profile
from flash_attn.io import read_QKV


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--T", type=int, default=8192)
    parser.add_argument("--d", type=int, default=64)
    parser.add_argument("--M_bytes", type=int, default=131072)
    parser.add_argument("--num_testsets", type=int, default=1)
    args = parser.parse_args()

    output_dir = os.path.join("outputs", "fa2_profile")
    os.makedirs(output_dir, exist_ok=True)
    log_path = os.path.join(output_dir, "runtime.csv")

    if not os.path.exists(log_path):
        with open(log_path, "w") as f:
            f.write("testset,T,d,M_bytes,runtime,line7,line8,line10,line11,others\n")

    completed_runs = set()
    with open(log_path, "r") as f:
        next(f, None)
        for line in f:
            parts = line.strip().split(",")
            if len(parts) >= 4:
                completed_runs.add((int(parts[0]), int(parts[1]),
                                    int(parts[2]), int(parts[3])))

    scale = np.float32(1.0 / np.sqrt(np.float32(args.d)))

    for i in range(args.num_testsets):
        key = (i, args.T, args.d, args.M_bytes)
        if key in completed_runs:
            print(f"Skipping testset{i}.")
            continue

        data_dir = os.path.join("data", f"testset{i}")
        Q, Kt, V = read_QKV(data_dir, args.T, args.d)

        _, runtime, t7, t8, t10, t11, others = fa2_forward_profile(
            Q, Kt, V, args.T, args.d, args.M_bytes, scale
        )

        with open(log_path, "a") as f:
            f.write(
                f"{i},{args.T},{args.d},{args.M_bytes},"
                f"{runtime:.6f},{t7:.6f},{t8:.6f},{t10:.6f},{t11:.6f},{others:.6f}\n"
            )

        print("-" * 30)
        print(
            f"Runtime {i}: {runtime:.6f} s\n"
            f"line7 {t7:.6f}s  line8 {t8:.6f}s  "
            f"line10 {t10:.6f}s  line11 {t11:.6f}s  others {others:.6f}s"
        )
        print("-" * 30)
